[draft][DNM] Add support for Nuclei E203 HummingBird RISC-V core

soburi opened the pull request:
This PR adds support for Nuclei E203 HummingBird RISC-V.

This PR depends on the external repository that proposing to add with previously open issue https://github.com/zephyrproject-rtos/zephyr/issues/39491.
(I also hope that this PR will help discuss that issue.)

It is only minimal to support. It can run only hello_world example.

fanghuaqi reviewed (commented):
Maybe `nuclei_e203hbird` could be renamed to `nuclei_hbird`， `hbird` is Nuclei's open source version of RISC-V CORE, it released two version:

1. hbird v1: https://github.com/SI-RISCV/e200_opensource
2. hbird v2: https://github.com/riscv-mcu/e203_hbirdv2/

The core insided it didn't change too much, but the integrated peripherals are changed, and the v2 version provded a lite version of NICE(customized instruction unit).

And the version you are integrating and pull requesting seems to be the hbird v1 version?

You can also check the support difference in our hbird_sdk for both version:

https://github.com/riscv-mcu/hbird-sdk/tree/master/SoC

fanghuaqi commented:
By the way, NMSIS in HBird SDK is a modified version to support PLIC interrupt controller which is not the ECLIC one used in normal NMSIS, maybe in future NMSIS could support both PLIC and ECLIC, but now not yet ready.

tokita-hiroshi committed:
soc: riscv: Add initial support of Nuclei E203 HummingBird

Add minimal support for Nuclei E203 HummingBird RISC-V core.
It includes minimal initialization code.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@jp.fujitsu.com>

tokita-hiroshi committed:
boards: riscv: Add Sipeed Lichee Tang

Sipeed Lichee Tang is Anlogic EG4S20 FPGA development board
shipped with Nuclei E203 HummingBird.

Add basic support for Lichee Tang with defalut E203
configuration.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@jp.fujitsu.com>

tokita-hiroshi committed:
drivers: serial: Add Nuclei E203 HummingBird UART

Introduce minimal serial driver support for E203 HummingBird.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@jp.fujitsu.com>

soburi commented:
I reworked the PR.

- Remove dependency to NMSIS. (Apply the result of discussion in https://github.com/zephyrproject-rtos/zephyr/issues/34971, HBird not require nuclei specific definitions.)
- Rename nuclei_e203hbird to hummingbird.
- Add a hierarchy (soc/riscv/riscv-privileged/nuclei/) to share Nuclei shared definitions.
- Place to soc/riscv/riscv-privileged/nuclei/hummingbird/

HBird implements CSR described in privileged ISA spec such as misa, mcause and others.
but HBird support only machine mode, not support supervisor mode and user mode.
My understand is that HBird is implementes privileged ISA, should be place under soc/riscv/riscv-privileged.

@carlescufi @fanghuaqi Could you confirm my understanding? Please.

At this time, west flash not correctly worked. I'm seeking workaround this now.

fkokosinski commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,142 @@
+
+
+/**
+ * @brief UART
```
Could you expand this comment? If I'm not mistaken, it's later extracted by Doxygen so `UART` description of this symbol is too short/generic.

soburi replied:
Adding a little more explain to comment.

fkokosinski commented on dts/bindings/serial/nuclei,hummingbird-uart.yaml:
```diff
@@ -0,0 +1,13 @@
+properties:
+    reg:
+      required: true
+
```
Please remove this blank line.

soburi replied:
Removed it.

fkokosinski commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,142 @@
+	volatile uint32_t IE;
+	volatile uint32_t IP;
+	volatile uint32_t DIV;
+} UART_TypeDef;
```
If I remember correctly, declaring new typedefs is against the Zephyr coding standard.

soburi replied:
Removed typedef and 
Replace UART_TypeDef with struct UART_TypeDef whole file.

fkokosinski commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,142 @@
+{
+	UART_TypeDef *uart = CFG2UART(cfg);
+
+	while (uart->TXFIFO & UART_TXFIFO_FULL);
```
Add an empty loop body here.

soburi replied:
Added empty brace pair.

soburi commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,142 @@
+
+#include <drivers/uart.h>
+
+#define _AC(X, Y)        (X##Y)
```
It can implements without _AC macro. Removing it.

soburi commented on soc/riscv/riscv-privilege/nuclei/hummingbird/entry.S:
```diff
@@ -0,0 +1,40 @@
+    csrsi mcountinhibit, 0x5
+
+    tail __initialize
+    //tail exit
```
Removed it.

soburi commented on soc/riscv/riscv-privilege/nuclei/linker.ld:
```diff
@@ -0,0 +1,400 @@
+#include <linker/linker-defs.h>
+#include <linker/linker-tool.h>
+
+
```
Removed it.

soburi commented on boards/riscv/lichee_tang/Kconfig.defconfig:
```diff
@@ -0,0 +1,10 @@
+# Sipeed Lichee Tang Development Board Configuration
+
+# Copyright (c) 2021 Tokita, Hiroshi <tokita.hiroshi@gmail.com>
```
@fkokosinski Thank you for reviewing.
I pushed the fix again. Could you check it again? please.
If the procedure is something wrong, please tell me.

gmarull commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,148 @@
+	struct hbird_uart_data *const data = dev->data;
+	struct UART_TypeDef *uart = CFG2UART(cfg);
+
+	GPIO_REG(GPIO_IOF_SEL) &= ~IOF0_UART0_MASK;
+	GPIO_REG(GPIO_IOF_EN) |= IOF0_UART0_MASK;
```
looks like `pinctrl` stuff?

gmarull commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,148 @@
+#define UART_IP_RXIP_OFS        (1)
+#define UART_IP_RXIP_MASK       (0x1 << UART_IP_RXIP_OFS)
+
+#define CFG2UART(cfg) ((struct UART_TypeDef *)((const struct hbird_uart_config *)dev->config)->reg)
```
please avoid macros like that: config can store a pointer to `UART_TypeDef`

gmarull commented on drivers/serial/uart_hummingbird.c:
```diff
@@ -0,0 +1,148 @@
+
+#include <drivers/uart.h>
+
+#define GPIO_CTRL_ADDR          0x10012000UL
+#define IOF0_UART0_MASK         0x00030000UL
+#define _REG32(p, i)            (*(volatile uint32_t *) ((p) + (i)))
+#define GPIO_REG(offset)        _REG32(GPIO_CTRL_ADDR, offset)
```
Afaik Zephyr has helper macros for register access, can't they be used?
